File: 0076-minimum-window-substring/0076-minimum-window-substring.py
from collections import Counter

class Solution:
    def minWindow(self, s: str, t: str) -> str:
        if not t or not s:
            return ""
        need = Counter(t)
        required = len(need)       
        have = 0                   
        window = {}
        start = 0
        minLength = float('inf')
        sub = ""
        for end in range(len(s)):
            c = s[end]
            window[c] = window.get(c, 0) + 1
            if c in need and window[c] == need[c]:
                have += 1
            while have == required:
                if end - start + 1 < minLength:
                    minLength = end - start + 1
                    sub = s[start:end + 1]
                left = s[start]
                window[left] -= 1
                if left in need and window[left] < need[left]:
                    have -= 1
                start += 1

        return sub

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (t.isEmpty() || s.isEmpty()) return "";

        Map<Character, Integer> need = new HashMap<>();
        for (char c : t.toCharArray()) {
            need.merge(c, 1, Integer::sum);
        }

        int required = need.size();   // distinct chars to satisfy
        int have = 0;

        Map<Character, Integer> window = new HashMap<>();
        int start = 0;
        int minLength = Integer.MAX_VALUE;
        int resStart = 0;             // remember where the best window began

        for (int end = 0; end < s.length(); end++) {
            char c = s.charAt(end);
            window.merge(c, 1, Integer::sum);

            if (need.containsKey(c) && window.get(c).intValue() == need.get(c).intValue()) {
                have++;
            }

            while (have == required) {
                if (end - start + 1 < minLength) {
                    minLength = end - start + 1;
                    resStart = start;
                }

                char left = s.charAt(start);
                window.merge(left, -1, Integer::sum);
                if (need.containsKey(left) && window.get(left) < need.get(left)) {
                    have--;
                }
                start++;
            }
        }

        return minLength == Integer.MAX_VALUE ? "" : s.substring(resStart, resStart + minLength);
    }
}